docs: Better documentation for developers (#427)
Safe-DS/Stdlib: docs/development/guidelines.md

    table.drop("name", axis="columns")
    ```

### Return copies of objects

Modifying objects in-place

> can lead to surprising behaviour

Reviewer: ```suggestion
can lead to surprising behaviours
```

Author: I have always thought of "behaviour" as an uncountable noun. I would also, for instance, say that "a program shows unexpected behaviour" (not "*an* unexpected behaviour").
Are you sure about this?

    training_feature_vectors, validation_feature_vectors, training_target_values, validation_target_values = split(feature_vectors, target_values)
    ```

> ## Code style

Reviewer 2: Better move this to the end of the document:

1. Design
2. Docstrings
3. Tests
4. Code style

Code style is the least important concern for our users, so it's also not as important for developers to keep in mind as the other items.

### Consistency

If there is more than one way
to solve a particular task,
check how it has been solved
at other places in the codebase
and stick to that solution.

### Sort exported classes in `__init__.py`

Classes defined in a module
that other classes shall be able to import
must be defined
in a list named `__all__`
in the module's `__init__.py` file.
This list should be sorted alphabetically,
to reduce the likelihood of merge conflicts
when adding new classes to it.

!!! success "**DO** (library code):"

    ```py
    __all__ = [

>     "Column",
>     "Row",
>     "Table",
>     "TaggedTable",

Reviewer 2: ```suggestion
        "Column",
        "Row",
        "Table",
        "TaggedTable",
```

    ]
    ```

!!! failure "**DON'T** (library code):"

    ```py
    __all__ = [

>     "Table",
>     "TaggedTable",
>     "Column",
>     "Row",

Reviewer 2: ```suggestion
        "Table",
        "TaggedTable",
        "Column",
        "Row",
```

> ## Tests

Reviewer 2: It might be worth splitting this into some more subsections, including:

* File structure (with do/don't example?)
* Naming (with do/don't example?)
* Parametrization (exists already)

Author: Splitted it into the suggested subsections and added a do / don't example for naming.

For the file structure, I'm not sure how such an example should look like. We cannot show Python code as in all the other examples.
At the same time, it should be fairly clear to understand, so I'm not sure examples are necessary here?

Reviewer 2: It's fine as it's now.

If a function contains more code
than just the getting or setting of a value,
automated test should be added

> to the [`tests`][tests-folder] folder.

Reviewer 2: A little outdated now. Maybe instead write something like

> We aim for 100% line coverage, so for any new function automated tests ...

Author: Changed it.

The file structure in the tests folder
should mirror the file structure
of the [`src`][src-folder] folder.

Names of test functions
shall start with `test_should_`
followed by a description
of the expected behaviour,
e.g. `test_should_add_column`.

> ### Testcases

Reviewer 2: ```suggestion
### Parametrization
```

Tests should be parametrized
using `@pytest.mark.parametrize`,

> even if there is only a single testcase.

Reviewer 2: Add reason: Easier to add new test cases in the future.

Author: Done.

> even if there is only a single testcase.

Reviewer 2: ```suggestion
even if there is only a single test case.
```

> Testcases should be given

Reviewer 2: ```suggestion
Test cases should be given
```

descriptive IDs.

Example:

    ```py
    @pytest.mark.parametrize(
        ("path", "expected"),
        [
            ("table.csv", Table({"A": [1], "B": [2]})),
            (Path("table.csv"), Table({"A": [1], "B": [2]})),
            ("empty_table.csv", Table()),
        ],
        ids=["by string", "by path", "empty"],
    )
    def test_should_create_table_from_csv_file(path: str | Path, expected: Table) -> None:
        ...

>     ```

Reviewer 2: Maybe have an example specifically for the parametrization and using the do/don't schema from above?
